File: internal/protocol/wtvb01/decoder_test.go
package wtvb01

import (
	"bufio"
	"encoding/hex"
	"math"
	"os"
	"strings"
	"testing"
)

// The fixtures in testdata/ are REAL bytes captured from a physical
// WitMotion WTVB01-BT50 (see the header of the .hex file). Synthetic
// packets are only used where a specific edge case cannot be captured
// on demand (resync, split writes), and are marked as such.

func loadCapture(t *testing.T) [][]byte {
	t.Helper()

	f, err := os.Open("testdata/capture-wtvb01-bt50.hex")
	if err != nil {
		t.Fatalf("open capture: %v", err)
	}
	defer f.Close()

	var payloads [][]byte
	scanner := bufio.NewScanner(f)
	scanner.Buffer(make([]byte, 0, 64*1024), 64*1024)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		b, err := hex.DecodeString(line)
		if err != nil {
			t.Fatalf("decode capture line: %v", err)
		}
		payloads = append(payloads, b)
	}
	if err := scanner.Err(); err != nil {
		t.Fatalf("read capture: %v", err)
	}
	if len(payloads) == 0 {
		t.Fatal("capture file contained no payloads")
	}
	return payloads
}

func almostEqual(a, b float64) bool { return math.Abs(a-b) < 1e-6 }

// TestDecodeRealCapture is the primary regression test: it replays the
// captured stream and checks the decoded values against what the sensor
// was physically doing (at rest on a desk, room temperature).
func TestDecodeRealCapture(t *testing.T) {
	d := NewDecoder()

	var last SensorReading
	decoded := 0
	for _, payload := range loadCapture(t) {
		reading, ok := d.Feed(payload)
		if ok {
			decoded++
			last = reading
		}
	}

	if decoded == 0 {
		t.Fatal("no packets decoded from the real capture")
	}

	// Chip temperature: the sensor sat in a room at roughly 24-25 C.
	// This is the one field confirmed against hardware.
	if last.Device.Temperature < 20 || last.Device.Temperature > 30 {
		t.Errorf("Device.Temperature = %v C, want a plausible ambient value (20-30)", last.Device.Temperature)
	}

	// At rest, vibration magnitudes must stay small and finite.
	for name, v := range map[string]Vector3{
		"Velocity":     last.Velocity,
		"Displacement": last.Displacement,
		"Frequency":    last.Frequency,
	} {
		for axis, val := range map[string]float64{"X": v.X, "Y": v.Y, "Z": v.Z} {
			if math.IsNaN(val) || math.IsInf(val, 0) {
				t.Errorf("%s.%s = %v, want a finite number", name, axis, val)
			}
		}
	}

	// Angle registers are scaled to degrees and must stay in range.
	for axis, val := range map[string]float64{"X": last.Angle.X, "Y": last.Angle.Y, "Z": last.Angle.Z} {
		if val < -180 || val > 180 {
			t.Errorf("Angle.%s = %v, want within [-180, 180]", axis, val)
		}
	}
}

// TestOutputAndRegisterPacketsAgree is the strongest correctness check
// available without the official app: the 0x61 broadcast and the 0x71
// register read-backs are independent encodings of the same registers,
// so decoding either must produce the same values.
func TestOutputAndRegisterPacketsAgree(t *testing.T) {
	payloads := loadCapture(t)

	// Split the captured stream into whole packets.
	var stream []byte
	for _, p := range payloads {
		stream = append(stream, p...)
	}

	var outputReading, registerReading SensorReading
	var gotOutput, gotRegisterA bool

	for i := 0; i+2 <= len(stream); {
		if stream[i] != syncByte {
			i++
			continue
		}
		want := packetLenFor(stream[i+1])
		if want == 0 || i+want > len(stream) {
			i++
			continue
		}
		pkt := stream[i : i+want]

		switch {
		case pkt[1] == packetTypeOutput && !gotOutput:
			d := NewDecoder()
			outputReading, gotOutput = d.Feed(pkt)
		case pkt[1] == packetTypeRegister && pkt[2] == blockMeasurementA && !gotRegisterA:
			d := NewDecoder()
			registerReading, gotRegisterA = d.Feed(pkt)
		}
		i += want
	}

	if !gotOutput || !gotRegisterA {
		t.Skip("capture did not contain both a 0x61 packet and a 0x3A register block")
	}

	// Block 0x3A covers velocity, angle and temperature. Values are
	// recomputed by the sensor continuously, so temperature drifts
	// between packets; the vibration registers update far more slowly
	// and must match exactly.
	if outputReading.Velocity != registerReading.Velocity {
		t.Errorf("Velocity mismatch: 0x61 %+v vs 0x71 %+v", outputReading.Velocity, registerReading.Velocity)
	}
	if outputReading.Angle != registerReading.Angle {
		t.Errorf("Angle mismatch: 0x61 %+v vs 0x71 %+v", outputReading.Angle, registerReading.Angle)
	}
	if math.Abs(outputReading.Device.Temperature-registerReading.Device.Temperature) > 1 {
		t.Errorf("Temperature mismatch: 0x61 %v vs 0x71 %v",
			outputReading.Device.Temperature, registerReading.Device.Temperature)
	}
}

// TestTemperatureScale pins the one scale confirmed against hardware:
// register 0x40 divided by 100 is degrees Celsius.
func TestTemperatureScale(t *testing.T) {
	// Real 0x71 read-back of block 0x3A from the capture. Register 0x40
	// (7th register, bytes 16-17) reads 0x0984 = 2436 -> 24.36 C.
	pkt, err := hex.DecodeString("55713a00110007001200e900f300390084092c01")
	if err != nil {
		t.Fatal(err)
	}

	d := NewDecoder()
	reading, ok := d.Feed(pkt)
	if !ok {
		t.Fatal("expected the register block to decode")
	}
	if !almostEqual(reading.Device.Temperature, 24.36) {
		t.Errorf("Device.Temperature = %v, want 24.36", reading.Device.Temperature)
	}
}

// TestPacketLengthsMatchHardware guards the correction that motivated
// this decoder: the WTVB01-BT50 broadcast is 32 bytes, not the 20 the
// generic BWT901 SDK assumes.
func TestPacketLengthsMatchHardware(t *testing.T) {
	if got := packetLenFor(packetTypeOutput); got != 32 {
		t.Errorf("packetLenFor(0x61) = %d, want 32", got)
	}
	if got := packetLenFor(packetTypeRegister); got != 20 {
		t.Errorf("packetLenFor(0x71) = %d, want 20", got)
	}
	if got := packetLenFor(0x99); got != 0 {
		t.Errorf("packetLenFor(unknown) = %d, want 0", got)
	}
}

// TestFeedResyncsOnGarbage uses a SYNTHETIC prefix: garbage ahead of a
// real captured packet, to prove the resync logic recovers.
func TestFeedResyncsOnGarbage(t *testing.T) {
	real, err := hex.DecodeString("55713a00110007001200e900f300390084092c01")
	if err != nil {
		t.Fatal(err)
	}
	garbage := []byte{0x00, 0xFF, 0x12, syncByte, 0x99} // 0x99 is not a valid type

	d := NewDecoder()
	reading, ok := d.Feed(append(garbage, real...))
	if !ok {
		t.Fatal("expected the decoder to resync and decode the trailing packet")
	}
	if !almostEqual(reading.Device.Temperature, 24.36) {
		t.Errorf("Device.Temperature = %v, want 24.36", reading.Device.Temperature)
	}
}

// TestFeedHandlesSplitWrites replays one real packet one byte at a
// time, since BLE notifications may be split across callbacks.
func TestFeedHandlesSplitWrites(t *testing.T) {
	pkt, err := hex.DecodeString("55713a00110007001200e900f300390084092c01")
	if err != nil {
		t.Fatal(err)
	}

	d := NewDecoder()
	var decoded bool
	var reading SensorReading
	for i := range len(pkt) {
		r, ok := d.Feed(pkt[i : i+1])
		if ok {
			decoded, reading = true, r
		}
	}

	if !decoded {
		t.Fatal("expected a packet after all bytes were fed")
	}
	if !almostEqual(reading.Device.Temperature, 24.36) {
		t.Errorf("Device.Temperature = %v, want 24.36", reading.Device.Temperature)
	}
}

File: firmware/esp32-zenith-node/test/decoder_test.cpp
// Host-side test for the WTVB01 decoder.
//
// Replays the same real captured sensor bytes the Go tests use, so both
// implementations are checked against identical input. Build and run:
//
//   c++ -std=c++17 -I src -o /tmp/wtvb01_test test/decoder_test.cpp src/wtvb01.cpp
//   /tmp/wtvb01_test ../../internal/protocol/wtvb01/testdata/capture-wtvb01-bt50.hex

#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <fstream>
#include <string>
#include <vector>

#include "wtvb01.h"

namespace {

int failures = 0;

void Check(bool ok, const char *what) {
  if (!ok) {
    std::printf("  FAIL: %s\n", what);
    failures++;
  }
}

std::vector<uint8_t> FromHex(const std::string &hex) {
  std::vector<uint8_t> out;
  for (size_t i = 0; i + 1 < hex.size(); i += 2) {
    out.push_back(
        static_cast<uint8_t>(std::stoul(hex.substr(i, 2), nullptr, 16)));
  }
  return out;
}

std::vector<std::vector<uint8_t>> LoadCapture(const char *path) {
  std::vector<std::vector<uint8_t>> payloads;
  std::ifstream f(path);
  if (!f) {
    std::printf("  FAIL: cannot open %s\n", path);
    failures++;
    return payloads;
  }
  std::string line;
  while (std::getline(f, line)) {
    while (!line.empty() && (line.back() == '\r' || line.back() == '\n')) {
      line.pop_back();
    }
    if (line.empty() || line[0] == '#') {
      continue;
    }
    payloads.push_back(FromHex(line));
  }
  return payloads;
}

void TestPacketLengths() {
  std::printf("TestPacketLengths\n");
  Check(wtvb01::PacketLenFor(wtvb01::kPacketTypeOutput) == 32,
        "0x61 packet is 32 bytes");
  Check(wtvb01::PacketLenFor(wtvb01::kPacketTypeRegister) == 20,
        "0x71 packet is 20 bytes");
  Check(wtvb01::PacketLenFor(0x99) == 0, "unknown type has no length");
}

void TestTemperatureScale() {
  std::printf("TestTemperatureScale\n");
  // Real 0x71 read-back of block 0x3A. Register 0x40 reads 0x0984 =
  // 2436 -> 24.36 C.
  const auto pkt = FromHex("55713a00110007001200e900f300390084092c01");

  wtvb01::Decoder d;
  Check(d.Feed(pkt.data(), pkt.size()), "register block decodes");
  Check(std::fabs(d.reading().device.temperature - 24.36f) < 0.01f,
        "temperature is 24.36 C");
}

void TestRealCapture(const char *path) {
  std::printf("TestRealCapture\n");
  const auto payloads = LoadCapture(path);
  if (payloads.empty()) {
    return;
  }

  wtvb01::Decoder d;
  int decoded = 0;
  for (const auto &p : payloads) {
    if (d.Feed(p.data(), p.size())) {
      decoded++;
    }
  }
  Check(decoded > 0, "packets decoded from the real capture");

  const auto &r = d.reading();
  std::printf("  last: vel(%.1f,%.1f,%.1f) disp(%.0f,%.0f,%.0f) "
              "ang(%.3f,%.3f,%.3f) freq(%.0f,%.0f,%.0f) temp=%.2f\n",
              r.velocity.x, r.velocity.y, r.velocity.z, r.displacement.x,
              r.displacement.y, r.displacement.z, r.angle.x, r.angle.y,
              r.angle.z, r.frequency.x, r.frequency.y, r.frequency.z,
              r.device.temperature);

  Check(r.device.temperature > 20.0f && r.device.temperature < 30.0f,
        "temperature is a plausible ambient value");
  Check(std::fabs(r.angle.x) <= 180.0f && std::fabs(r.angle.y) <= 180.0f &&
            std::fabs(r.angle.z) <= 180.0f,
        "angles within [-180, 180]");
}

// The 0x61 broadcast and the 0x71 read-backs independently encode the
// same registers, so decoding either must agree.
void TestOutputAndRegisterAgree(const char *path) {
  std::printf("TestOutputAndRegisterAgree\n");
  const auto payloads = LoadCapture(path);
  if (payloads.empty()) {
    return;
  }

  std::vector<uint8_t> stream;
  for (const auto &p : payloads) {
    stream.insert(stream.end(), p.begin(), p.end());
  }

  wtvb01::SensorReading from_output, from_register;
  bool got_output = false, got_register = false;

  for (size_t i = 0; i + 2 <= stream.size();) {
    if (stream[i] != wtvb01::kSyncByte) {
      i++;
      continue;
    }
    const size_t want = wtvb01::PacketLenFor(stream[i + 1]);
    if (want == 0 || i + want > stream.size()) {
      i++;
      continue;
    }

    if (stream[i + 1] == wtvb01::kPacketTypeOutput && !got_output) {
      wtvb01::Decoder d;
      if (d.Feed(&stream[i], want)) {
        from_output = d.reading();
        got_output = true;
      }
    } else if (stream[i + 1] == wtvb01::kPacketTypeRegister &&
               stream[i + 2] == wtvb01::kRegVelocityX && !got_register) {
      wtvb01::Decoder d;
      if (d.Feed(&stream[i], want)) {
        from_register = d.reading();
        got_register = true;
      }
    }
    i += want;
  }

  if (!got_output || !got_register) {
    std::printf("  SKIP: capture lacked both packet types\n");
    return;
  }

  Check(from_output.velocity.x == from_register.velocity.x &&
            from_output.velocity.y == from_register.velocity.y &&
            from_output.velocity.z == from_register.velocity.z,
        "velocity agrees between 0x61 and 0x71");
  Check(from_output.angle.x == from_register.angle.x &&
            from_output.angle.y == from_register.angle.y &&
            from_output.angle.z == from_register.angle.z,
        "angle agrees between 0x61 and 0x71");
  // Temperature is recomputed continuously, so it drifts slightly.
  Check(std::fabs(from_output.device.temperature -
                  from_register.device.temperature) < 1.0f,
        "temperature agrees between 0x61 and 0x71");
}

void TestResyncOnGarbage() {
  std::printf("TestResyncOnGarbage\n");
  const auto real = FromHex("55713a00110007001200e900f300390084092c01");
  std::vector<uint8_t> stream = {0x00, 0xFF, 0x12, wtvb01::kSyncByte, 0x99};
  stream.insert(stream.end(), real.begin(), real.end());

  wtvb01::Decoder d;
  Check(d.Feed(stream.data(), stream.size()), "decoder resyncs after garbage");
  Check(std::fabs(d.reading().device.temperature - 24.36f) < 0.01f,
        "temperature correct after resync");
}

void TestSplitWrites() {
  std::printf("TestSplitWrites\n");
  const auto pkt = FromHex("55713a00110007001200e900f300390084092c01");

  wtvb01::Decoder d;
  bool decoded = false;
  for (size_t i = 0; i < pkt.size(); i++) {
    if (d.Feed(&pkt[i], 1)) {
      decoded = true;
    }
  }
  Check(decoded, "packet decoded across single-byte writes");
  Check(std::fabs(d.reading().device.temperature - 24.36f) < 0.01f,
        "temperature correct after split writes");
}

}  // namespace

int main(int argc, char **argv) {
  const char *capture =
      argc > 1 ? argv[1]
               : "../../internal/protocol/wtvb01/testdata/"
                 "capture-wtvb01-bt50.hex";

  TestPacketLengths();
  TestTemperatureScale();
  TestRealCapture(capture);
  TestOutputAndRegisterAgree(capture);
  TestResyncOnGarbage();
  TestSplitWrites();

  if (failures > 0) {
    std::printf("\n%d check(s) FAILED\n", failures);
    return 1;
  }
  std::printf("\nall checks passed\n");
  return 0;
}
